enclave/seal.c

```diff
@@ -0,0 +1,121 @@
+// Copyright (c) Open Enclave SDK contributors.
+// Licensed under the MIT License.
+
+#include <openenclave/enclave.h>
+#include <openenclave/corelibc/stdlib.h>
+#include <openenclave/internal/crypto/gcm.h>
+#include <openenclave/internal/raise.h>
+#include <openenclave/internal/safecrt.h>
+#include <openenclave/internal/entropy.h>
+
+static const uint8_t _iv[12] = { 0 };
+
+oe_result_t oe_seal(
+    const uint8_t* key_info,
+    size_t key_info_size,
+    const uint8_t* plaintext,
+    size_t plaintext_size,
+    const uint8_t* additional_data,
```

Comment:
Exposing additional data implies that OE SDK is now signing up to offer authenticated encryption that supports additional data, which seems to run counter to the expectation TEEs will want to define the crypto suited that works best for them (per the given example TPM).

It's also not clear to me that offering the capability to provide integrity-protected but unencrypted data in the same function helps developers simplify the sealing experience; it's another thing they can mess up (e.g. providing a blob that should be encrypted, rather than just integrity protected), especially if different TEEs are permitted to handle the data in different ways.

Reply:
Thank @CodeMonkeyLeet for your good points! Technically, `additional_data` is meant to be accessed without having to decrypt the blob. But the blob format here is private, and the only way to retrieve `additional_data` is to decrypt the blob via `oe_unseal()`, which violates the intention of `additional_data`. I can remove it from the API. The question then will become the interoperability with Intel SGX SDK. Do you think interoperability is important/desirable?